### ISSUE TITLE
Circe: Allow to decode enumerations more relaxed

### ISSUE THREAD

**Issue by Fristi** [lloydmeta/enumeratum#392]
A pattern I've been using lately is 

```scala

sealed trait MyEnum extends EnumEntry

object MyEnum extends Enum[MyEnum] {
  case object A extends MyEnum
  case object B extends MyEnum
  case class Unknown(other: String) extends MyEnum

  override def values =
    findValues
}

```

Later for circe encoders/decoders

```scala
implicit protected val codecMyEnum: Codec[MyEnum] =
    Codec.from(fallbackDecoder(MyEnum, MyEnum.Unknown), Circe.encoder(MyEnum))

  def fallbackDecoder[A <: EnumEntry](e: Enum[A], ifNotFound: String => A) =
    Decoder.decodeString.map { s =>
      e.withNameOption(s) match {
        case Some(member) => member
        case None         => ifNotFound(s)
      }
    }

```

The rationale behind is that some code which uses that matches the exact case and in case it doesn'tt recognize it will continue. With explicitly defined enumerations the code might break easily, which is undesired in a environment with API's changing quickly and not all consumers should be up-to-date all the time, it's a high maintenance burden.

Bottom line, would you be open to contribute `fallbackDecoder` (likely with a case insenstive version?)

**Comment by lloydmeta** [comment]
Thanks for the issue; while I understand the use case, right now, I'm open to discussion but not sold on the proposed solution:

1. It involves having an `Unknown` member of _each_ such enum, which is its own boilerplate and exhaustive-pattern-checking tax/burden.
2. It would involve poking holes in the `findValues` macro to handle (ignore?) the new dynamic member(s).

IMO, maybe a more natural solution for your usecase might be to have your API-level decoders be union-based (Scala 3, or Coproducts via Shapeless), like

```scala
// new class for holding unknowns, reusable for all instances where you want 
// to support ok-to-be-unknown
final case class Unknown(s: String)

// This is now just your normal enum
sealed trait MyEnum extends EnumEntry

object MyEnum extends Enum[MyEnum] {
  case object A extends MyEnum
  case object B extends MyEnum

  override val values =  findValues
}


// Union type decoder
def fallbackToUnknownDecoder[A <: EnumEntry](e: Enum[A]): Enum[A] | Unknown =
  Decoder.decodeString.map { s =>
    e.withNameOption(s) match {
      case Some(member) => member
      case None         => Unknown(s)
    }
  }
```

You can also probably do something without union/coproducts with some kind of a wrapper type:

```scala
sealed trait EnumOrUnknown[A <: EnumEntry]

final case class HasEnum[A <: EnumEntry](enum : Enum[A]) extends EnumOrUnknown[A]
final case class Unknown[A <: EnumEntry](s: String) extends EnumOrUnknown[A]

object EnumOrUnknown {
  def decoder[A <: EnumEntry](e: Enum[A]):  EnumOrUnknown[A] =
    Decoder.decodeString.map { s =>
      e.withNameOption(s) match {
        case Some(member) => EnumOrUnknown (member)
        case None         => Unknown(s)
      }
    }
}
```

**Comment by Fristi** [comment]
Agreed on all your points, unfortunately we don't have Scala 3 yet :)

I liked the latter solution for Scala 2 the most, keeping it most natural :) 

I guess none of this would be beneficial to end up in this lib or does it ?

**Comment by lloydmeta** [comment]
It's an interesting use case but not something I think I'd like to maintain in this lib, so I would encourage you to either publish a supporting lib or keep it in your apps 🙂
